pkg/utils/controller/controller_utils.go

```diff
@@ -61,3 +72,213 @@ func CreateHeadlessServiceIfNotExists(ctx context.Context, k8sClient client.Clie
+// newRevision creates a new ControllerRevision containing a patch that reapplies the target state of LeaderWorkerSet.
+// The Revision of the returned ControllerRevision is set to revision. If the returned error is nil, the returned
+// ControllerRevision is valid. LeaderWorkerSet revisions are stored as patches that re-apply the current state of set
+// to a new LeaderWorkerSet using a strategic merge patch to replace the saved state of the new LeaderWorkerSet.
+func NewRevision(lws *leaderworkerset.LeaderWorkerSet, revision int64, templateHash string) (*appsv1.ControllerRevision, error) {
+	patch, err := getPatch(lws)
+	if err != nil {
+		return nil, err
+	}
+	cr, err := history.NewControllerRevision(lws,
+		controllerKind,
+		map[string]string{leaderworkerset.TemplateRevisionHashKey: templateHash},
+		runtime.RawExtension{Raw: patch},
+		revision)
+	if err != nil {
+		return nil, err
+	}
+	if cr.ObjectMeta.Annotations == nil {
+		cr.ObjectMeta.Annotations = make(map[string]string)
+	}
+	for key, value := range lws.Annotations {
+		cr.ObjectMeta.Annotations[key] = value
+	}
+	return cr, nil
+}
+
+// ApplyRevision returns a new LeaderWorkerSet constructed by restoring the state in revision to set. If the returned error
+// is nil, the returned LeaderWorkerSet is valid.
+func ApplyRevision(lws *leaderworkerset.LeaderWorkerSet, revision *appsv1.ControllerRevision) (*leaderworkerset.LeaderWorkerSet, error) {
+	clone := lws.DeepCopy()
+	str := &bytes.Buffer{}
+	err := unstructured.UnstructuredJSONScheme.Encode(lws, str)
+	if err != nil {
+		return nil, err
+	}
+	patched, err := strategicpatch.StrategicMergePatch(str.Bytes(), revision.Data.Raw, clone)
+	if err != nil {
+		return nil, err
+	}
+	restoredLws := &leaderworkerset.LeaderWorkerSet{}
+	err = json.Unmarshal(patched, restoredLws)
+	if err != nil {
+		return nil, err
+	}
+	return restoredLws, nil
+}
+
+// nextRevision finds the next valid revision number based on revisions. If the length of revisions
+// is 0 this is 1. Otherwise, it is 1 greater than the largest revision's Revision. This method
+// assumes that revisions has been sorted by Revision.
+func NextRevision(revisions []*appsv1.ControllerRevision) int64 {
+	count := len(revisions)
+	if count <= 0 {
+		return 1
+	}
+	return revisions[count-1].Revision + 1
+}
+
+// TruncateHistory cleans up all other controller revisions except the currentRevision.
+// currentRevision is the one that matches the templateHash that is passed
+func TruncateHistory(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) error {
+	controllerHistory := history.NewHistory(ctx, k8sClient)
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{})
+	if err != nil {
+		return err
+	}
+	revisions, err := controllerHistory.ListControllerRevisions(lws, selector)
+	if err != nil {
+		return err
+	}
+	currentRevision, err := GetLeaderWorkerSetRevisionFromTemplateHash(ctx, k8sClient, lws, templateHash)
```

Comment:
No need for this, truncate all revision that has a different templateHash, no need to get it and have the condition on the name as line 277 does.